### ISSUE TITLE
Maximum call stack size exceeded when cloning a path with many segments

### ISSUE THREAD

**Issue by arthursw** [paperjs/paper.js#1493]
# Description/Steps to reproduce

 1. Create a path with more than 125122 segments
 2. Clone it

# Link to reproduction test-case

http://sketch.paperjs.org/#S/VY7BDoIwEER/ZdPEAIEQwXgBOfkDJh48iIdCV9sUuwSqHAj/bku8sKfZmczbnZnhb2QFu2q0rWQJa0n4vUMLPbcSKjA4wcXJMKqNt9LRDqTxTB0NLg6ajrc6KDfhTYm1mzn7SUPocaraQwnqlOXHLD94GccRzLUBN2uXCxGq3VfhlDb0MWJMJ89JYGtKVC9pI4de/kfbjgyGUen+bwbkuidl7MiK+2P5AQ==

# Note

It is possible to get around this problem by splitting the path before cloning it.

**Comment by lehni** [comment]
Hmm that's strange. It works for me on Chrome. What system / browser are you experiencing this on?

**Comment by arthursw** [comment]
Sorry I should have given my specs right away.

I used chrome Version 67.0.3396.99 (Build officiel) (64 bits) 
on OS X High Seria Version 10.13.6 (17G65)

Strangely, the number of segments to reach stack overflow increased by one on Chomre.
On Firefox, I could clone much more segments, but at 1 000 000 I got: 

`Line 8118: RangeError: too many arguments provided for a function call`

Anyway, thanks a million time again for creating Paper.js! 
I use it for the Tipibot [controller](https://arthursw.github.io/tipibot-controller-gui/)  ( [Comme un Dessein](http://idlv.co/comme-un-dessein/), [tipibot](http://idlv.co/tipibot/), [robograff](http://idlv.co/robograff/), [IDLV](http://idlv.co/)), me and my friends are greatly inspired by Viktor and Hektor (and polargraph) !

**Comment by lehni** [comment]
Interesting. I can reproduce that too now. It looks like `segments.push.apply(segments, segs)` isn't such a good idea :)

maybe it should be `segments = segments.concat(segs)` instead.

**Comment by sasensi** [comment]
@lehni, I prefered using `.push()` in a loop over `.concat()` because it mutates existing array instead of copying it and thus seems faster.
Here is the PR #1534.

**Comment by sasensi** [comment]
As a side note to this issue, the same kind of issue will eventually come out by calling `Path.insert()` instead of `Path.add()`.
But I don't see what optimization we could do for this one as it relies on `Array.splice()` and calling it in a loop won't be much better.
https://github.com/paperjs/paper.js/blob/53897d749819b86bc47e935294e14971323a21e9/src/path/Path.js#L410-L414

**Comment by lehni** [comment]
@sasensi it's hard to say which one is faster... `push()` needs to grow the allocated memory for the array without knowing how much it'l grow, `concat()` can do that once and then just add.  To avoid the allocation issue, we could set `segments.length` to the new length before the loop, this will grow it once. Could you add that?

**Comment by sasensi** [comment]
@lehni, when I was saying that it seemed faster, I was not aware of the internal difference between `push()` and `concat()`, I just saw a neat difference by trying both with the 1000000 segments case in the browser. `concat()` was taking forever and `push()` only a few seconds.
I implemented the allocation optimization [here](https://github.com/paperjs/paper.js/pull/1534/commits/d12b99e252114a204c7d000f9045e9f8cc864380).
